### PR TITLE
Refresh AccessKey passively

### PR DESCRIPTION
**Pull request by terencefan**

### CODE REVIEW

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs [hunk 1]
```diff
@@ -119,55 +117,61 @@ public async Task<string> GenerateAccessTokenAsync(string audience,
                                                        AccessTokenAlgorithm algorithm,
                                                        CancellationToken ctoken = default)
     {
-        if (!_initializedTcs.Task.IsCompleted)
+        var updateTask = Task.CompletedTask;
+        if (NeedRefresh)
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:121]: NeedRefresh is false before Available is set, so it is possible that multiple `cts` are created, move the cts creation into the lock?

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs [hunk 2]
```diff
@@ -119,55 +117,61 @@ public async Task<string> GenerateAccessTokenAsync(string audience,
...
+        lock (_lock)
         {
-            return;
+            if (!_updateTaskSource.Task.IsCompleted)
+            {
+                source = _updateTaskSource;
+            }
+            else
+            {
+                _updateTaskSource = new TaskCompletionSource<bool>();
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs]: `TaskCreationOptions.RunContinuationsAsynchronously`

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs [hunk 3]
```diff
@@ -119,55 +117,61 @@ public async Task<string> GenerateAccessTokenAsync(string audience,
...
+            _updateTaskSource.TrySetResult(true);
         }
+    }
 
-        if (Interlocked.CompareExchange(ref _updateState, UpdateTaskRunning, UpdateTaskIdle) != UpdateTaskIdle)
+    internal async Task<bool> UpdateAccessKeyAsync(CancellationToken ctoken = default)
+    {
+        TaskCompletionSource<bool> source;
+        lock (_lock)
         {
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:157]: double check NeedRefresh inside the lock

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs [hunk 4]
```diff
@@ -119,55 +117,61 @@ public async Task<string> GenerateAccessTokenAsync(string audience,
...
-            return;
-        }
-        else if (!Available && delta < GetAccessKeyIntervalWhenUnauthorized)
+        lock (_lock)
         {
-            return;
+            _keyBytes = Encoding.UTF8.GetBytes(keyStr);
+            _kid = kid;
+            Available = true;
+            _updateTaskSource.TrySetResult(true);
```
Comment (vicancy) [inline src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:149]: only _updateTaskSource needs to be in the lock?

File: src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs (hunk 1 above)
Comment (terencefan) [reply src/Microsoft.Azure.SignalR.Common/Auth/MicrosoftEntra/MicrosoftEntraAccessKey.cs:121]: NeedRefresh is always true, the init value of `_updatedAt` is `DateTime.MinValue`

**Review by vicancy — APPROVED** [review]

**Pull request merged**